File: bindings/nodejs/examples/example-basic.js
#!/usr/bin/env node
/**
 * Basic Elecrypto usage examples for Node.js
 */

const elecrypto = require('../dist/index');

console.log('=== Elecrypto Node.js Bindings Examples ===\n');

// 1. Hash functions
console.log('1. Hash Functions');
const message = Buffer.from('Hello, Elecrypto!');

const sha256Hash = elecrypto.sha256(message);
console.log(`SHA-256: ${sha256Hash.toString('hex')}`);

const sha512Hash = elecrypto.sha512(message);
console.log(`SHA-512: ${sha512Hash.toString('hex').slice(0, 64)}...`);

const blake3Hash = elecrypto.blake3(message);
console.log(`BLAKE3:  ${blake3Hash.toString('hex')}\n`);

// 2. Random number generation
console.log('2. Random Number Generation');
const randomData = elecrypto.randomBytes(32);
console.log(`32 random bytes: ${randomData.toString('hex')}\n`);

// 3. AES-GCM encryption
console.log('3. AES-256-GCM Encryption');
const plaintext = Buffer.from('This is a secret message!');
const aesKey = elecrypto.aesGenerateKey();
console.log(`Generated key: ${aesKey.toString('hex')}`);

const { ciphertext: aesCiphertext, nonce: aesNonce } = elecrypto.aesGcmEncrypt(plaintext, aesKey);
console.log(`Ciphertext: ${aesCiphertext.toString('hex')}`);
console.log(`Nonce: ${aesNonce.toString('hex')}`);

const aesDecrypted = elecrypto.aesGcmDecrypt(aesCiphertext, aesKey, aesNonce);
console.log(`Decrypted: ${aesDecrypted.toString()}`);
console.log('✓ Encryption/Decryption successful\n');

// 4. ChaCha20-Poly1305 encryption
console.log('4. ChaCha20-Poly1305 Encryption');
const chachaKey = elecrypto.chacha20GenerateKey();
const { ciphertext: chachaCiphertext, nonce: chachaNonce } = elecrypto.chacha20Poly1305Encrypt(plaintext, chachaKey);
console.log(`Ciphertext: ${chachaCiphertext.toString('hex')}`);

const chachaDecrypted = elecrypto.chacha20Poly1305Decrypt(chachaCiphertext, chachaKey, chachaNonce);
console.log(`Decrypted: ${chachaDecrypted.toString()}`);
console.log('✓ ChaCha20 encryption successful\n');

// 5. Digital signatures (Ed25519)
console.log('5. Ed25519 Digital Signatures');
const { publicKey, secretKey } = elecrypto.ed25519GenerateKeypair();
console.log(`Public key: ${publicKey.toString('hex')}`);
console.log(`Secret key: ${secretKey.toString('hex').slice(0, 32)}...`);

const signMessage = Buffer.from('Sign this important message');
const signature = elecrypto.ed25519Sign(signMessage, secretKey);
console.log(`Signature: ${signature.toString('hex').slice(0, 32)}...`);

const isValid = elecrypto.ed25519Verify(signMessage, signature, publicKey);
console.log(`Signature valid: ${isValid}`);

// Try with wrong message
const wrongMessage = Buffer.from('Different message');
const isValidWrong = elecrypto.ed25519Verify(wrongMessage, signature, publicKey);
console.log(`Wrong message valid: ${isValidWrong}`);
console.log('✓ Digital signature successful\n');

// 6. Key derivation (PBKDF2)
console.log('6. Key Derivation - PBKDF2');
const password = Buffer.from('super_secret_password');
const pbkdf2Salt = elecrypto.randomBytes(16);

const pbkdf2Key = elecrypto.pbkdf2(password, pbkdf2Salt, 100000, 32);
console.log(`Derived key: ${pbkdf2Key.toString('hex')}`);
console.log('✓ PBKDF2 successful\n');

// 7. Key derivation (Argon2id)
console.log('7. Key Derivation - Argon2id');
const { derivedKey, salt } = elecrypto.argon2id(password, undefined, 65536, 3, 4, 32);
console.log(`Derived key: ${derivedKey.toString('hex')}`);
console.log(`Salt: ${salt.toString('hex')}`);
console.log('✓ Argon2id successful\n');

// 8. HKDF
console.log('8. HKDF Key Derivation');
const inputKey = elecrypto.randomBytes(32);
const info = Buffer.from('application context');

const hkdfKey = elecrypto.hkdf(inputKey, undefined, info, 64);
console.log(`Derived key (64 bytes): ${hkdfKey.toString('hex')}`);
console.log('✓ HKDF successful\n');

console.log('=== All tests passed! ===');

File: bindings/python/examples/example_basic.py
#!/usr/bin/env python3
"""
Basic Elecrypto usage examples
"""

import sys
import os

# Add parent directory to path for development
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

import elecrypto

print("=== Elecrypto Python Bindings Examples ===\n")

# 1. Hash functions
print("1. Hash Functions")
message = b"Hello, Elecrypto!"

sha256_hash = elecrypto.sha256(message)
print(f"SHA-256: {sha256_hash.hex()}")

sha512_hash = elecrypto.sha512(message)
print(f"SHA-512: {sha512_hash.hex()[:64]}...")

blake3_hash = elecrypto.blake3(message)
print(f"BLAKE3:  {blake3_hash.hex()}\n")

# 2. Random number generation
print("2. Random Number Generation")
random_data = elecrypto.random_bytes(32)
print(f"32 random bytes: {random_data.hex()}\n")

# 3. AES-GCM encryption
print("3. AES-256-GCM Encryption")
plaintext = b"This is a secret message!"
key = elecrypto.aes_generate_key()
print(f"Generated key: {key.hex()}")

ciphertext, nonce = elecrypto.aes_gcm_encrypt(plaintext, key)
print(f"Ciphertext: {ciphertext.hex()}")
print(f"Nonce: {nonce.hex()}")

decrypted = elecrypto.aes_gcm_decrypt(ciphertext, key, nonce)
print(f"Decrypted: {decrypted.decode()}")
assert decrypted == plaintext, "Decryption failed!"
print("✓ Encryption/Decryption successful\n")

# 4. ChaCha20-Poly1305 encryption
print("4. ChaCha20-Poly1305 Encryption")
key = elecrypto.chacha20_generate_key()
ciphertext, nonce = elecrypto.chacha20_poly1305_encrypt(plaintext, key)
print(f"Ciphertext: {ciphertext.hex()}")

decrypted = elecrypto.chacha20_poly1305_decrypt(ciphertext, key, nonce)
print(f"Decrypted: {decrypted.decode()}")
assert decrypted == plaintext
print("✓ ChaCha20 encryption successful\n")

# 5. Digital signatures (Ed25519)
print("5. Ed25519 Digital Signatures")
public_key, secret_key = elecrypto.ed25519_generate_keypair()
print(f"Public key: {public_key.hex()}")
print(f"Secret key: {secret_key.hex()[:32]}...")

message = b"Sign this important message"
signature = elecrypto.ed25519_sign(message, secret_key)
print(f"Signature: {signature.hex()[:32]}...")

is_valid = elecrypto.ed25519_verify(message, signature, public_key)
print(f"Signature valid: {is_valid}")
assert is_valid, "Signature verification failed!"

# Try with wrong message
wrong_message = b"Different message"
is_valid = elecrypto.ed25519_verify(wrong_message, signature, public_key)
print(f"Wrong message valid: {is_valid}")
assert not is_valid
print("✓ Digital signature successful\n")

# 6. Key derivation (PBKDF2)
print("6. Key Derivation - PBKDF2")
password = b"super_secret_password"
salt = elecrypto.random_bytes(16)

derived_key = elecrypto.pbkdf2(password, salt, iterations=100000, key_length=32)
print(f"Derived key: {derived_key.hex()}")
print("✓ PBKDF2 successful\n")

# 7. Key derivation (Argon2id)
print("7. Key Derivation - Argon2id")
derived_key, salt = elecrypto.argon2id(
    password,
    memory_cost=65536,  # 64 MB
    time_cost=3,
    parallelism=4,
    key_length=32
)
print(f"Derived key: {derived_key.hex()}")
print(f"Salt: {salt.hex()}")
print("✓ Argon2id successful\n")

# 8. HKDF
print("8. HKDF Key Derivation")
input_key = elecrypto.random_bytes(32)
info = b"application context"

derived = elecrypto.hkdf(input_key, info=info, key_length=64)
print(f"Derived key (64 bytes): {derived.hex()}")
print("✓ HKDF successful\n")

print("=== All tests passed! ===")
